### PR TITLE
feat: Add ability to filter GCP project ingestion based on project labels

### PR DESCRIPTION
**Pull request by anaghshineh**
# What Changed?

We've wanted to have more control over project ingestion at my organization. We've been maintaining a "whitelist" of projects we'd like to ingest, but this is becoming unmanageable as we create more projects. We'd like to leverage consistent project labels in GCP to drive ingestion. This PR takes a first pass at allowing folks to define `project_labels` in their BigQuery ingestion recipes. It uses the GCP Resource Manager `search_projects` method to query for projects with the given labels. This does not introduce any breaking changes, as ingestion will continue to favor `project_ids` & will respect any name patterns.

Example:

```
source:
    type: bigquery
    config:
        project_labels:
            - "team:data"
            - "domain:*"
        credential:
            project_id: '${BIGQUERY_INGESTION_PROJECT}'
            client_email: '${SERVICE_ACCOUNT_CLIENT_EMAIL}'
            private_key_id: '${BIGQUERY_PRIVATE_KEY_ID}'
            private_key: '${BIGQUERY_PRIVATE_KEY}'
            client_id: '${SERVICE_ACCOUNT_UNIQUE_ID}'
        project_id_pattern:
            deny:
                - '^sys-[0-9]{26}$'
sink:
    type: datahub-rest
    config:
        server: '${DATAHUB_BASE_URL}'
        token: '${DATAHUB_ACCESS_TOKEN}'
        max_threads: 20
```

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_config.py [hunk 1]
```diff
@@ -69,6 +69,9 @@ def __init__(self, **data: Any):
     def get_bigquery_client(self) -> bigquery.Client:
         client_options = self.extra_client_options
         return bigquery.Client(self.project_on_behalf, **client_options)
+    
+    def get_projects_client(self) -> resourcemanager_v3.ProjectsClient:
+        return resourcemanager_v3.ProjectsClient()
```
Comment (treff7es) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_config.py:74]: Do you know if this client needs extra permission, or does it not need any new permission?
If it is needed, then can we make sure not to create this client if, in the config, it is not set to use it?

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_config.py [hunk 2]
```diff
@@ -193,6 +196,13 @@ class BigQueryV2Config(
             "Overrides `project_id_pattern`."
         ),
     )
+    project_labels: List[str] = Field(
+        default_factory=list,
+        description=(
+            "Ingests projects with the specified labels. Use this properly if you want to specify what "
```
Comment (treff7es) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_config.py:202]:
```suggestion
            "Ingests projects with the specified labels. Use this property if you want to specify what "
```

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_config.py (hunk 1 above)
Comment (anaghshineh) [reply metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_config.py:74]: Good question. I initially implemented this PR used the `list_projects` method but landed on the `search_projects` one due to its flexibility in searching for projects based on parents AND labels. Looking at [GCP docs](https://cloud.google.com/resource-manager/reference/rest/v3/projects/search), seems like you'll need to ensure the SA has `resourcemanager.projects.get` on the projects you want to grab.

Is that your read, too?

File: docker/datahub-ingestion-base/base-requirements.txt [hunk 3]
```diff
@@ -1,16 +1,17 @@
 # Generated requirements file. Run ./regenerate-base-requirements.sh to regenerate.
```
Comment (anshbansal) [inline docker/datahub-ingestion-base/base-requirements.txt:1]: Please don't make changes in this file. This file is updated separately. This is mainly a caching mechanism for our builds.

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_config.py (hunk 1 above)
Comment (anshbansal) [reply metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_config.py:74]: @anaghshineh There is a easy way to test this out by creating a service account without this permission and try to run this code. If this does indeed require a new permission then please
- update the docs with the new roles/permissions required
- make this an optional thing disabled by default. We don't want every orgs ingestion to break due to this new permission. Some orgs have heavier checks on getting new permissions. We don't want them to get stuck on permissions when this gets in. 
- Please add in https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md that there is new option that requires new permissions

As this is open source we have to be careful w.r.t. not breaking other orgs ingestion as much as possible.

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery.py [hunk 4]
```diff
@@ -337,10 +339,12 @@ def metadata_read_capability_test(
         for project_id in project_ids:
             try:
                 logger.info((f"Metadata read capability test for project {project_id}"))
-                client: bigquery.Client = config.get_bigquery_client()
-                assert client
+                bq_client: bigquery.Client = config.get_bigquery_client()
+                assert bq_client
+                proj_client: resourcemanager_v3.Client = config.get_projects_client()
+                assert proj_client
```
Comment (anshbansal) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery.py:345]: Please put this under a config disabled by default. We don't want folks ingestions to break due to new permissions required as mentioned in other comment.

**Comment by hsheth2** [comment]
Closing as https://github.com/datahub-project/datahub/pull/11169 was merged

**Pull request closed without merging**